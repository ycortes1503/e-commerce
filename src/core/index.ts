export * from './models';
export { Router } from './Router';
export * from './routes';
