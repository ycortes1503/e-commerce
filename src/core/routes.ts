export const ROUTES = {
  ROOT: '/',
  HOME: '/home',
  DETAILS: '/details',
  CHECKOUT: '/checkout',
};
