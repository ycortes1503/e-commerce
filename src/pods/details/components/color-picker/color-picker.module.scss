@use '@/assets/scss/variables/colors' as *;

.container {
  display: flex;
  gap: 16px;
  flex-direction: column;

  &__items-container {
    display: flex;
    gap: 16px;
  }
}

.item {
  width: 24px;
  height: 24px;
  border: 1px solid $border-grey;
  box-sizing: border-box;

  &:hover {
    border: 1px solid $black;
    scale: 1.1;
    cursor: pointer;
  }

  &--selected {
    border: 1px solid $black;
  }
}

.item-color {
  width: 100%;
  height: 100%;
  border: 1px solid $white;
  box-sizing: border-box;
}
