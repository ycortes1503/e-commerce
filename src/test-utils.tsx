import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { CartProvider } from './context';
import { MemoryRouter as Router } from 'react-router-dom';
const WrapperProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <CartProvider>
      <Router>{children}</Router>
    </CartProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: WrapperProviders, ...options });

export { customRender as render };

import { MemoryRouter } from 'react-router-dom';

const renderWithRouter = (ui, { route = '/' } = {}) => {
  return render(<MemoryRouter initialEntries={[route]}>{ui}</MemoryRouter>);
};

export { renderWithRouter };

export * from '@testing-library/react';
