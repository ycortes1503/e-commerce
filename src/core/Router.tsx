import * as React from 'react';
import { Routes, Route } from 'react-router-dom';
import { LazyCheckoutScene, LazyDetailsScene, LazyHomeScene, LazyNotFoundScene } from '../scenes';
import { ROUTES } from './routes';

export const Router: React.FC = () => {
  return (
    <React.Suspense fallback={<div>Loading...</div>}>
      <Routes>
        <Route path={ROUTES.ROOT} element={<LazyHomeScene />} />
        <Route path={ROUTES.HOME} element={<LazyHomeScene />} />
        <Route path={`${ROUTES.DETAILS}/:id`} element={<LazyDetailsScene />} />
        <Route path={ROUTES.CHECKOUT} element={<LazyCheckoutScene />} />
        <Route path="*" element={<LazyNotFoundScene />} />
      </Routes>
    </React.Suspense>
  );
};
