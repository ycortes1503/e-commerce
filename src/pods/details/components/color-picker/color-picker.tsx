import * as React from 'react';
import cx from 'classnames';
import * as styles from './color-picker.module.scss';
import { ColorOption } from '@/core';
import { isNullOrUndefined } from '@/helpers';
import { Typography } from '@/common/components';

export interface ColorPickerProps {
  colorOptions: ColorOption[];
  onSelect: (option: ColorOption) => void;
}

export const ColorPicker = ({ colorOptions, onSelect }: ColorPickerProps) => {
  const defaultColor = colorOptions[0];
  const [selectedColor, setSelectedColor] = React.useState<ColorOption>(defaultColor);
  const [colorHovering, setColorHovering] = React.useState<ColorOption>(null);

  const handleSelect = (option: ColorOption) => {
    setSelectedColor(option);
    onSelect(option);
  };

  const colorLabel = !isNullOrUndefined(colorHovering) ? colorHovering.name : selectedColor.name;

  return (
    <div className={styles.container}>
      <div className={styles['container__items-container']}>
        {colorOptions.map((option, index) => {
          const key = `${index}-${option.name}`;

          return (
            <div
              key={key}
              data-testid={`color-option-${index}`}
              className={cx(styles['item'], {
                [styles['item--selected']]: selectedColor.name === option.name && isNullOrUndefined(colorHovering),
              })}
              onClick={() => handleSelect(option)}
              onMouseOver={() => setColorHovering(option)}
              onMouseOut={() => setColorHovering(null)}
            >
              <div className={styles['item-color']} style={{ backgroundColor: option.hexCode }} />
            </div>
          );
        })}
      </div>
      <Typography fontSize="12" text={colorLabel} />
    </div>
  );
};
