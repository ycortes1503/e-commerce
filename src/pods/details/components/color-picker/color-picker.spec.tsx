import { fireEvent, render, screen } from '@/test-utils';
import { ColorPicker } from './color-picker';
import { ColorOption } from '@/core';

describe('<ColorPicker />', () => {
  const colorOptions: ColorOption[] = [
    { name: 'red', hexCode: '#ff0000', imageUrl: 'irrelevant imageUrl' },
    { name: 'blue', hexCode: '#0000ff', imageUrl: 'irrelevant imageUrl' },
  ];

  const mockFunction = jest.fn();

  it('should render correctly', () => {
    const { container } = render(<ColorPicker colorOptions={colorOptions} onSelect={mockFunction} />);

    expect(container).toMatchSnapshot();
  });

  it('should call onSelect when a color option is selected', () => {
    render(<ColorPicker colorOptions={colorOptions} onSelect={mockFunction} />);

    const colorOption = screen.getByTestId('color-option-0');
    fireEvent.click(colorOption);

    expect(mockFunction).toHaveBeenCalledWith(colorOptions[0]);
  });
});
