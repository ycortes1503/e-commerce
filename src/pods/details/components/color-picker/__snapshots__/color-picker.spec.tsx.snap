// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker /> should render correctly 1`] = `
<div>
  <div>
    <div>
      <div
        class="undefined"
        data-testid="color-option-0"
      >
        <div
          style="background-color: rgb(255, 0, 0);"
        />
      </div>
      <div
        class=""
        data-testid="color-option-1"
      >
        <div
          style="background-color: rgb(0, 0, 255);"
        />
      </div>
    </div>
    <p
      class=""
      style="font-size: 12px;"
    >
      red
    </p>
  </div>
</div>
`;
